### PR TITLE
Sc sfmlearner yaak efficientloader train inference

### PR DESCRIPTION
**Pull request by arturomg82**
Added new features:

- Integrated the efficient data loader which reads configuration files (TOML).
- Added a speed-based regularizer to the loss function (to constrain the output of the pose network) during training.
- Implemented a script to perform model inference. Additional data is computed/stored per frame (validation losses, estimated position/orientation between two frames, speed, latitude, longitude, Tensorboard logs, etc).

### CODE REVIEW

File: inverse_warp.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 
 def set_id_grid(depth):
+
```
Comment (Anaaa) [inline inverse_warp.py:9]: Probably Black gonna remove the empty line

File: inverse_warp.py [hunk 2]
```diff
@@ -85,54 +61,99 @@ def euler2mat(angle):
...
     """Convert quaternion coefficients to rotation matrix.
     Args:
         quat: first three coeff of quaternion of rotation. fourht is then computed to have a norm of 1 -- size = [B, 3]
     Returns:
         Rotation matrix corresponding to the quaternion -- size = [B, 3, 3]
     """
-    norm_quat = torch.cat([quat[:, :1].detach()*0 + 1, quat], dim=1)
+
+    # norm_quat = torch.cat([quat[:, :1].detach()*0 + 1, quat], dim=1)
```
Comment (Anaaa) [inline inverse_warp.py:132]: Remove?

File: inverse_warp.py [hunk 3]
```diff
@@ -203,31 +191,37 @@ def cam2pixel2(cam_coords, proj_c2p_rot, proj_c2p_tr, padding_mode):
     b, _, h, w = cam_coords.size()
     cam_coords_flat = cam_coords.reshape(b, 3, -1)  # [B, 3, H*W]
     if proj_c2p_rot is not None:
-        pcoords = proj_c2p_rot @ cam_coords_flat
+        # pcoords = proj_c2p_rot @ cam_coords_flat
```
Comment (Anaaa) [inline inverse_warp.py:194]: Remove?

File: inverse_warp.py [hunk 4]
```diff
@@ -250,20 +244,24 @@ def inverse_warp2(img, depth, ref_depth, pose, intrinsics, padding_mode='zeros')
 
     batch_size, _, img_height, img_width = img.size()
 
-    cam_coords = pixel2cam(depth.squeeze(1), intrinsics.inverse())  # [B,3,H,W]
+    # cam_coords = pixel2cam(depth.squeeze(1), intrinsics.inverse())  # [B,3,H,W]
```
Comment (Anaaa) [inline inverse_warp.py:247]: Remove?

File: loss_functions.py [hunk 5]
```diff
@@ -42,97 +49,566 @@ def forward(self, x, y):
...
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # Initialization.
+    # ------------------------------------------------------------------------------------------------------------------
+
+    # If convert_speed_kmph_to_mps is True, the vehicle's speed is converted from kilometers per hour (mk/h)
+    # to meters per second (m/s). Otherwise, the current speed is kept (in km/h).
+    scaled_speed_data = \
+        (1000. * speed_data.unsqueeze(-1)) / (60. ** 2) \
+        if convert_speed_kmph_to_mps else speed_data.unsqueeze(-1)
```
Comment (Anaaa) [inline loss_functions.py:107]: Please use parenthesis instead of backslashes. This is the preferred python way (PEP8, also Black will replace `\` with `(...)`)

File: loss_functions.py [hunk 6]
```diff
@@ -42,97 +49,566 @@ def forward(self, x, y):
...
 
     photo_loss = 0
     geometry_loss = 0
+    velocity_supervision_loss = 0
 
     num_scales = min(len(tgt_depth), max_scales)
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # Loop over the data...
+    # ------------------------------------------------------------------------------------------------------------------
```
Comment (Anaaa) [inline loss_functions.py:245]: Not needed in my opinion

File: loss_functions.py [hunk 7]
```diff
@@ -42,97 +49,566 @@ def forward(self, x, y):
...
+
+    # Index for reference images.
+    ref_img_idx = 0
+
     for ref_img, ref_depth, pose, pose_inv in zip(ref_imgs, ref_depths, poses, poses_inv):
+
+        # --------------------------------------------------------------------------------------------------------------
+        # Loop over scales.
+        # --------------------------------------------------------------------------------------------------------------
```
Comment (Anaaa) [inline loss_functions.py:254]: Not needed in my opinion

File: loss_functions.py [hunk 8]
```diff
@@ -42,97 +49,566 @@ def forward(self, x, y):
...
+            photo_loss2, geometry_loss2, vel_superv_loss2 = \
+                compute_pairwise_loss(
+                    ref_img_scaled, tgt_img_scaled,
+                    ref_depth_scaled, tgt_depth_scaled,
+                    pose_inv, intrinsic_scaled,
+                    with_ssim, with_mask, with_auto_mask,
+                    padding_mode,
+                    rotation_mode=rotation_mode,
+                    velocity_supervision_loss_params=velocity_supervision_loss_params,
+                    writer_obj_tag='{}_pairwise_loss_target_wrt_ref{}_reversed_scale_{}'.format(
```
Comment (Anaaa) [inline loss_functions.py:326]: f-string?

File: loss_functions.py [hunk 9]
```diff
@@ -42,97 +49,566 @@ def forward(self, x, y):
...
+                        writer_obj_tag, ref_img_idx, s
+                    ),
+                    writer_obj_step=writer_obj_step,
+                    writer_obj=writer_obj,
+                    device=device,
+                )
+
+            # Photometric loss.
```
Comment (Anaaa) [inline loss_functions.py:334]: When you call a variable just like this, then no extra comment is needed.

File: loss_functions.py [hunk 10]
```diff
@@ -42,97 +49,566 @@ def forward(self, x, y):
...
             photo_loss += (photo_loss1 + photo_loss2)
+
+            # Geometry consistency loss.
             geometry_loss += (geometry_loss1 + geometry_loss2)
 
-    return photo_loss, geometry_loss
+            # Velocity supervision loss.
```
Comment (Anaaa) [inline loss_functions.py:340]: Remove, the name of the variable is self-explainging

File: loss_functions.py [hunk 11]
```diff
@@ -42,97 +49,566 @@ def forward(self, x, y):
...
+
+        ref_img_idx += 1
+
+    return photo_loss, geometry_loss, velocity_supervision_loss
+
+########################################################################################################################
+#
+# Pair-wise loss.
+#
+########################################################################################################################
```
Comment (Anaaa) [inline loss_functions.py:356]: Remove?

File: loss_functions.py [hunk 12]
```diff
@@ -42,97 +49,566 @@ def forward(self, x, y):
...
 
-    if with_ssim == True:
-        ssim_map = compute_ssim_loss(tgt_img, ref_img_warped)
+        # Log data...
+        if writer_obj is not None:
+            writer_obj.add_image(
+                tag='{}/auto_mask'.format(writer_obj_tag),
+                img_tensor=valid_mask[0],
+                global_step=writer_obj_step
+            )
```
Comment (Anaaa) [inline loss_functions.py:554]: If `valid_mask = auto_mask` could you use only one variable, e.g. `valid_mask`?

File: loss_functions.py [hunk 13]
```diff
@@ -158,32 +634,56 @@ def get_smooth_loss(disp, img):
...
+#
+########################################################################################################################
+
+
+def mean_on_mask(diff, valid_mask, device=torch.device("cpu")):
+
+    """ Compute mean value given a binary mask. """
+
+    mask = valid_mask.expand_as(diff)
+    if mask.sum() > 10000:
```
Comment (Anaaa) [inline loss_functions.py:649]: This looks strange, why?

File: models/DispResNet.py [hunk 14]
```diff
@@ -83,37 +95,106 @@ def __init__(self, num_ch_enc, scales=range(4), num_output_channels=1, use_skips
...
 
     def forward(self, input_features):
+
         self.outputs = []
 
         # decoder
         x = input_features[-1]
+
+        if self.verbose:
+            print('[ DepthDecoder ][ Input ] x.shape = ', x.shape)
```
Comment (Anaaa) [inline models/DispResNet.py:105]: `logger.info` is rather a better choice. With loggers you can set up the level of logs, play with hierarchy, etc. Also, when you use `logger.debug` level, it may be a powerful tool for debugging.

File: models/DispResNet.py [hunk 15]
```diff
@@ -83,37 +95,106 @@ def __init__(self, num_ch_enc, scales=range(4), num_output_channels=1, use_skips
...
             x = [upsample(x)]
+
+            if self.verbose:
+                print('\t\t[ upsample(x) ] Shape = {}'.format(x[0].shape))
+
             if self.use_skips and i > 0:
                 x += [input_features[i - 1]]
+
+            if self.verbose:
+                print('\t\t[ input_features[{:d}] ] Shape = {}'.format(i-1, input_features[i-1].shape))
```
Comment (Anaaa) [inline models/DispResNet.py:127]: when `i==0` then you try to access `input_features[-1]` is this correct?

File: models/DispResNet.py [hunk 16]
```diff
@@ -83,37 +95,106 @@ def __init__(self, num_ch_enc, scales=range(4), num_output_channels=1, use_skips
...
-        self.decoder = DepthDecoder(self.encoder.num_ch_enc)
+
+        # Encoder network.
+        self.encoder = ResnetEncoder(
+            num_layers=num_layers,
+            pretrained=pretrained,
+            num_input_images=1
+        )
+
+        # Decoded network.
```
Comment (Anaaa) [inline models/DispResNet.py:160]: Here also

File: models/DispResNet.py [hunk 17]
```diff
@@ -83,37 +95,106 @@ def __init__(self, num_ch_enc, scales=range(4), num_output_channels=1, use_skips
...
+        self.decoder = DepthDecoder(
+            num_ch_enc=self.encoder.num_ch_enc,
+            scales=range(4),
+            num_output_channels=1,
+            use_skips=True,
+            verbose=verbose,
+        )
+
+        # Verbose.
```
Comment (Anaaa) [inline models/DispResNet.py:169]: and here

File: run_inference.py [hunk 18]
```diff
@@ -1,83 +1,1377 @@
-import torch
-
-from imageio import imread, imsave
-from scipy.misc import imresize
-import numpy as np
-from path import Path
+import traceback
+import sys
+import shutil
+sys.path.insert(1, '/home/arturo/workspace/pycharm_projects/data_loader_ml/DataLoaderML')
```
Comment (Anaaa) [inline run_inference.py:4]: No, this is really a bad idea :P Basically, anybody else gonna fail, when you change the location - fail. Please use git submodules. A better option would be to make dataloader installable package but we are not there yet.

File: run_inference.py [hunk 19]
```diff
@@ -1,83 +1,1377 @@
...
+
+    print(f"[ Data Loader {YaakIterableDataset.desc()}, Version {str(YaakIterableDataset.version())} ]")
+    print(" ")
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # Global vars.
+    # ------------------------------------------------------------------------------------------------------------------
+
+    # global best_error, n_iter, device
+    global best_error, n_iter
```
Comment (Anaaa) [inline run_inference.py:146]: Why do you need global variables?

File: run_inference.py [hunk 20]
```diff
@@ -1,83 +1,1377 @@
...
+    # ------------------------------------------------------------------------------------------------------------------
+
+    print(f"[ Data Loader {YaakIterableDataset.desc()}, Version {str(YaakIterableDataset.version())} ]")
+    print(" ")
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # Global vars.
+    # ------------------------------------------------------------------------------------------------------------------
+
+    # global best_error, n_iter, device
```
Comment (Anaaa) [inline run_inference.py:145]: Remove? device is outdated

File: run_inference.py [hunk 21]
```diff
@@ -1,83 +1,1377 @@
...
+        print(f'[ Experimental results ][ {index+1} of {N} ] Drive ID = {current_drive_id}')
+        print(f'\t- Save path: {save_path}')
+
+        # If the path does not exist, it is created.
+        if not os.path.exists(save_path):
+            os.makedirs(save_path)
+            print(f'\t- Creating directory: {save_path}')
+
+        # Check the value of the arguments pretrained_disp and pretrained_pose.
+        pretrained_disparity_model_file = None if cfgs["experiment_settings"]["pretrained_disp"] == "None" \
```
Comment (Anaaa) [inline run_inference.py:204]: `cfgs["experiment_settings"]["pretrained_disp"] == "None"` don't the load function translate to `None`? The should be a value that is translated to `None` (could be `null` or a special flag)

File: run_inference.py [hunk 22]
```diff
@@ -1,83 +1,1377 @@
...
+
+            print('\n[ Initializing Weights & Biases ]')
+            print('\t- Project name: {}'.format(wandb_project_name))
+            print('\t- Metadata path: {}'.format(wandb_dir))
+            print('\t- Tensorboard root log dir: {}'.format(tensorboard_root_logdir))
+            print(' ')
+
+        # --------------------------------------------------------------------------------------------------------------
+        # Summary writers...
+        # --------------------------------------------------------------------------------------------------------------
```
Comment (Anaaa) [inline run_inference.py:266]: Instead of addind comments that indicate a block of code, why not to extract the code to a separate function?

File: run_inference.py [hunk 23]
```diff
@@ -1,83 +1,1377 @@
...
+                    source_video_ffname,
+                    # outputdict={'-vcodec': 'libx264', '-b': '300000000'}
+                )
+
+                if verbose:
+                    print("[ Video writer ] Object created.")
+                    print(f"\t- RGB and disparity frames will be saved in: {source_video_ffname}")
+                    print(" ")
+
+            except Exception as err:
```
Comment (Anaaa) [inline run_inference.py:716]: Is it possible to catch sth more precise?

File: run_inference.py [hunk 24]
```diff
@@ -1,83 +1,1377 @@
...
+                print(f"[ Exception ] Error while creating an video writer object with skvideo.io.FFmpegWriter: {err}")
+                traceback.print_exc()
```
Comment (Anaaa) [inline run_inference.py:718]: Should the program continue?

File: utils.py [hunk 25]
```diff
@@ -40,27 +42,324 @@ def opencv_rainbow(resolution=1000):
...
+        Here, n_i and w_i stand for the i-th number of iterations and the corresponding weight factor, respectively.
+        For instance, "10:0.01; 10:0.02; 10:0.03; 10:0.04; 10:0.05; 200:0.05".
+
+    """
+
+    if isinstance(value, float) or isinstance(value, int):
+
+        return value
+
+    elif isinstance(value, str):
```
Comment (Anaaa) [inline utils.py:347]: The conditions are incomplete, if value is neither of float, int nor string, it returns silently None. I would either replace this elif with else or add at the end `raise NotImplemented`

**Review by sandhawalia — CHANGES_REQUESTED** [review]
Please lint the code with `black` and add instructions on how to run inference
